Terminal.Gui/Windows/FileDialog2.cs

```diff
@@ -0,0 +1,2118 @@
+			/// </summary>
+			/// <param name="description">The human readable text to display.</param>
+			/// <param name="extensions">Extension(s) to match e.g. .csv.</param>
+			public AllowedType (string description, params string [] extensions)
+			{
+				if (extensions.Length == 0) {
+					throw new ArgumentException ("You must supply at least one extension");
+				}
+
+				this.Description = description;
+				this.Extensions = extensions;
+			}
+
+			/// <summary>
+			/// Gets a value of <see cref="AllowedType"/> that matches any file.
+			/// </summary>
+			public static AllowedType Any { get; } = new AllowedType ("Any Files", ".*");
+
+			/// <summary>
+			/// Gets or Sets the human readable description for the file type
+			/// e.g. "Comma Separated Values".
+			/// </summary>
+			public string Description { get; set; }
+
+			/// <summary>
+			/// Gets or Sets the permitted file extension(s) (e.g. ".csv").
+			/// </summary>
+			public string [] Extensions { get; set; }
+
+			/// <summary>
+			/// Gets a value indicating whether this instance is the
+			/// static <see cref="Any"/> value which indicates matching
+			/// any files.
+			/// </summary>
+			public bool IsAny => this == Any;
+
+			/// <summary>
+			/// Returns <see cref="Description"/> plus all <see cref="Extensions"/> separated by semicolons.
+			/// </summary>
+			public override string ToString ()
+			{
+				return $"{this.Description} ({string.Join (";", this.Extensions.Select (e => '*' + e).ToArray ())})";
+			}
+
+			internal bool Matches (string extension, bool strict)
+			{
+				if (this.IsAny) {
+					return !strict;
+				}
+
+				return this.Extensions.Any (e => e.Equals (extension));
+			}
+		}
+
+
+		/// <summary>
+		/// Defines whether a given file/directory matches a set of
+		/// search terms.
+		/// </summary>
+		public interface ISearchMatcher 
+		{
+			/// <summary>
+			/// Called once for each new search.  Defines the string
+			/// the user has provided as search terms.
+			/// </summary>
+			void Initialize(string terms);
+
+			/// <summary>
+			/// Return true if <paramref name="f"/> is a match to the
+			/// last provided search terms
+			/// </summary>
+			bool IsMatch(FileSystemInfo f);
+		}
+
+		class DefaultSearchMatcher : ISearchMatcher
+		{
+			string terms;
+
+			public void Initialize (string terms)
+			{
+				this.terms = terms;
+			}
+
+			public bool IsMatch (FileSystemInfo f)
+			{
+				//Contains overload with StringComparison is not available in (net472) or (netstandard2.0)
+				//return f.Name.Contains (terms, StringComparison.OrdinalIgnoreCase);
+
+				// This is the same
+				return f.Name.IndexOf (terms, StringComparison.OrdinalIgnoreCase) >= 0;
+			}
+		}
+
+		/// <summary>
+		/// Wrapper for <see cref="FileSystemInfo"/> that contains additional information
+		/// (e.g. <see cref="IsParent"/>) and helper methods.
+		/// </summary>
+		internal class FileSystemInfoStats {
```

Comment:
Did you consider using extension methods instead of a wrapper? I'm not sure it's a good idea, I am just curious.

Reply:
Definetly needs to be a wrapper.  The advantages are:

- Persist stuff needed for repaint (e.g. human readable size) so it is not constantly recalculated
- Allows for special instances of the class e.g. `..` for navigating up one level

Its internal so we can always change the implementation later on if theres a problem but I feel like this is the best way to go for now.